cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,484 @@
+		// If node was modified within this delta, delete this change.
+		delete(data.modifiedNodeInfoMap, nodeName)
+	}
+
+	_, foundInBase := data.baseData.getNodeInfo(nodeName)
+	if foundInBase {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if !foundInBase && !foundInDelta {
+		// Node not found in the chain.
+		return errNodeNotFound
+	}
+
+	// Maybe consider deleting from the lists instead. Maybe not.
+	data.clearCaches()
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	dni.AddPod(pod)
+	if data.podList != nil || data.havePodsWithAffinity != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) removePod(namespace, name, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	preAffinityPods := len(dni.PodsWithAffinity())
+
+	podFound := false
+	for _, pod := range dni.Pods() {
+		if pod.Namespace == namespace && pod.Name == name {
+			if err := dni.RemovePod(pod); err != nil {
+				return fmt.Errorf("cannot remove pod; %v", err)
+			}
+			podFound = true
+			break
+		}
+	}
+	if !podFound {
+		return fmt.Errorf("pod %s/%s not in snapshot", namespace, name)
+	}
+
+	// Maybe consider deleting from the list in the future. Maybe not.
+	postAffinityPods := len(dni.PodsWithAffinity())
+	if preAffinityPods == 1 && postAffinityPods == 0 {
+		data.havePodsWithAffinity = nil
+	}
+	if data.podList != nil {
+		data.clearPodCaches()
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) getPodList() []*apiv1.Pod {
+	if data == nil {
+		return nil
+	}
+	if data.podList == nil {
+		data.podList = data.buildPodList()
+	}
+	return data.podList
+}
+
+func (data *internalDeltaSnapshotData) buildPodList() []*apiv1.Pod {
+	podList := []*apiv1.Pod{}
+	if len(data.deletedNodeInfos) > 0 || len(data.modifiedNodeInfoMap) > 0 {
+		nodeInfos := data.getNodeInfoList()
+		for _, ni := range nodeInfos {
+			podList = append(podList, ni.Pods()...)
+		}
+	} else {
+		basePodList := data.baseData.getPodList()
+		copy(podList, basePodList)
+		for _, ni := range data.addedNodeInfoMap {
+			podList = append(podList, ni.Pods()...)
+		}
+	}
+	return podList
+}
+
+func (data *internalDeltaSnapshotData) fork() *internalDeltaSnapshotData {
+	forkedData := newInternalDeltaSnapshotData()
+	forkedData.baseData = data
+	return forkedData
+}
+
+func (data *internalDeltaSnapshotData) commit() *internalDeltaSnapshotData {
```

Comment:
This is way more tricky than it seems once you start following the implementation of the helper methods. I would very much like to see a test that does all sorts of modifications on fork (add/delete pod on pre-existing nodes, add/delete pod on pre-existing node and later delete this node) and verifies that everything is applied correctly on commit.